text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+---
+feature name: cli-framework-compatibility-strategy
+start date: February 2nd, 2020
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/111
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/110
+---
+
+<!-- replace the blockquoted sections with your content -->
+# Summary
+
+This RFC addresses compatibility concerns for developing different parts of the CDK.
+
+It details exactly what compatibility promises we should be advertising to our customers,
+as well as proposes technical mechanisms for validating those promises are upheld.
+
+At a high level, we define the expected customer experience when upgrading either of the two CDK components:
+
+- CLI upgrades are compatible. New CLI versions should work with older framework versions, and all existing functionality is preserved.
```

Comment:
I think what you are referring to is part of framework integ tests, which are out of scope for this RFC. Any issues that CLI code can cause should be covered by CLI integ tests, which is essentially what i mean by "functionality is preserved". 

Do you think just a rephrasing is in order or is there a validation mechanism missing?

Reply:
I need to know what "functionality" means here. I can break the hell out of the API (just renaming all operations), while retaining all of the functionality. But this is technically nuclear-scale breakage.

And yeah, there's the thing about framework interactions, which is a gray area here. That actually makes the scope of this statement ambiguous and it would benefit rephrasing for sure.

Reply:
>  I can break the hell out of the API (just renaming all operations), while retaining all of the functionality

I think the word `API` is actually misleading. We should talk about `schema`, you wouldn't be able to break the `schema` (lets say rename properties), without breaking cli functionality with regards to older framework versions, because the CLI would expect non-existing properties. I guess my point is that `functionality` encompasses the entire CLI experience. The functionality itself can be broken because of many reasons, but at the end of the day, the user simply expects the cli to "Work". 

I'll try to come up with some clarification but i don't really feel this is as ambiguous as you claim, especially from a user point of view.

Reply:
BTW, if you are somehow able to break the hell out of the schema, and the templates, and whatever...but still make the CLI work with older framework versions, that's arguably fine. 

Anyway, i've added a small block quote to try and clarify, but i'm considering this discussion resolved for now. Feel free to re-open :)